x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -113,7 +124,13 @@ public class Analyzer extends ParameterizedRuleExecutor<LogicalPlan, AnalyzerCon
             new ResolveFunctions(),
             new RemoveDuplicateProjections()
         );
-        var finish = new Batch<>("Finish Analysis", Limiter.ONCE, new AddImplicitLimit(), new PromoteStringsInDateComparisons());
+        var finish = new Batch<>(
+            "Finish Analysis",
+            Limiter.ONCE,
+            new AddImplicitLimit(),
+            new PromoteStringsInDateComparisons(),
```

Comment:
> So, maybe I misunderstood from our conversation last week, but I thought we wanted the to do the auto-casting either before `ResolveFunctions()` or as part of `ResolveFunctions()`. My initial read of this is that this code will never get called, because the incompatible type will cause the query to be rejected as invalid while we're resolving the functions.

Yeah, in our original discussion, the auto-casting is part of ResolveFunctions(). The resolveType() of each function is called by analyzer's verifier, it is at the very end of analyzer, after all of those resolutions, it is possible that a function is "resolved" by ResolveFunctions(), but fails at verifier because data type mismatches. The auto-casting can be done somewhere between ResolveFunctions and analyzer's verifier, as long as the data type is correct when verifier checks it. One advantage of having a separate rule for it (that I can think of) is that it makes it easier to be extended to other situations like BinaryComparison, ArithmeticOperation, or move it to logical optimizer.

Reply:
Let's double check this scenario by using a function as a child to other expressions that might be resolved.
something like:  `from test | eval x = date_trunc(1 month, "2024-11-22") % 2, y = x/3 | stats count() by y`

Reply:
> Let's double check this scenario by using a function as a child to other expressions that might be resolved. something like: `from test | eval x = date_trunc(1 month, "2024-11-22") % 2, y = x/3 | stats count() by y`

This query fails on both main and branch with different errors:
from main:
1:24: first argument of [date_trunc(1 month, \"2024-11-22\")] must be [datetime], found value [\"2024-11-22\"] type [keyword]\nline 1:67: Unknown column [x]

from branch:
1:24: first argument of [date_trunc(1 month, \"2024-11-22\") % 2] must be [numeric], found value [date_trunc(1 month, \"2024-11-22\")] type [datetime]\nline 1:67: Unknown column [x]

The actual evaluation of date_trunc happens in logical optimizer, x is not a literal value at analyzer phase, so it is as expected. More changes need to be done(in both analyzer and logical optimizer) if we want to make this query succeed.

Reply:
Implicit casting for arithmetic operations is added in the last commit. The ImplicitCasting rule is moved inside the resolution batch, as the type resolutions of some more complicated queries need to apply this rule multiple times to be able to resolve the expressions referenced in the query.

These queries failed on main:
from employees | eval x = languages + "1", y = x * "2";
second argument of [languages + "1"] must be [datetime or numeric], found value ["1"] type [keyword]

from employees | eval x = date_trunc(1 month, "2024-11-22") + 2 days, y = x + 3 days | sort emp_no | keep emp_no, x, y | limit 2;
first argument of [date_trunc(1 month, "2024-11-22")] must be [datetime], found value ["2024-11-22"] type [keyword]

from employees | eval x = date_trunc(1 month, "2024-11-22") + 2 days, y = x + 3 days | stats count() by y;
first argument of [date_trunc(1 month, "2024-11-22")] must be [datetime], found value ["2024-11-22"] type [keyword]

After supporting literal casting for arithmetic operations, they all succeed.

I'll do more testing while people are reviewing.